File: Leetcode/15_reverse_linked_list.py
# https://leetcode.com/problems/reverse-linked-list/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# 반복 방식
def reverseList(head: ListNode) -> ListNode:
    prev, current = None, head

    while current:
        current.next, prev, current = prev, current, current.next

    return prev


# 재귀 방식
# def reverseList(head: ListNode) -> ListNode:
#     def reverse(current: ListNode, prev: ListNode = None):
#         if not current:
#             return prev

#         current.next, next = prev, current.next

#         return reverse(next, current)

#     return reverse(head)


node1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)
node4 = ListNode(4)
node5 = ListNode(5)

node1.next = node2
node2.next = node3
node3.next = node4
node4.next = node5

print(reverseList(node1).val)

File: Leetcode/15_reverse_linked_list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * 재귀로 풀기
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  // 리스트를 순회하는 재귀 함수
  function reverse(current, prev = null) {
    // 리스트 순회가 끝났으므로 가장 마지막 노드를 리턴한다.
    if (!current) return prev;

    // 노드를 역순으로 뒤집는다. (prev <- current)
    [current.next, prev, current] = [prev, current, current.next];

    // 재귀 호출
    return reverse(current, prev);
  }

  return reverse(head);
};

/**
 * 반복으로 풀기
 * @param {ListNode} head
 * @return {ListNode}
 */
// var reverseList = function (head) {
//   let [prev, current] = [null, head];

//   while (current) {
//     [current.next, current, prev] = [prev, current.next, current];
//   }

//   return prev;
// };
